mempool/clist_mempool.go

```diff
@@ -253,7 +255,23 @@ func (mem *CListMempool) CheckTx(tx types.Tx, cb func(*abci.Response), txInfo Tx
 		}
 	}
 
-	// CACHE
+	// NOTE: writing to the WAL and calling proxy must be done before adding tx
+	// to the cache. otherwise, if either of them fails, next time CheckTx is
+	// called with tx, ErrTxInCache will be returned without tx being checked at
+	// all even once.
+	if mem.wal != nil {
+		// TODO: Notify administrators when WAL fails
+		_, err := mem.wal.Write(append([]byte(tx), newline...))
+		if err != nil {
+			return fmt.Errorf("wal.Write: %w", err)
```

Comment:
It might still be beneficial to keep the error log here.

Reply:
the error will be logged by the caller